File: Lectures/Object-Oriented/04-Inheritance-and-Function-Overriding.py
class BaseClass:
    pass
class DerivedClass(BaseClass):
    pass


class Human:
    def __init__(self, name, age):
        self.name = name
        self.age = age
    def __str__(self):
        return 'name: ' + self.name + '\n' + \
               'age: ' + str(self.age) + '\n'

class Employee(Human):
    def __init__(self, name, age, salary, department):
        Human.__init__(self, name, age)
        self.salary = salary
        self.department = department
    def __str__(self):
        return Human.__str__(self) + \
               'salary: ' + str(self.salary) + '\n' + \
               'department: ' + self.department + '\n'

x = Human('Jack', 24)
print(x)

y = Employee('Robert', 26, 4000, 'IT')
print(y)

#    name: Jack
#    age: 24

#    name: Robert
#    age: 26
#    salary: 4000
#    department: IT


class Human:
    def __init__(self, name, age):
        self.name = name
        self.age = age
    def __str__(self):
        return 'name: ' + self.name + '\n' + \
               'age: ' + str(self.age) + '\n'

class Employee(Human):
    def __init__(self, name, age, salary, department):
        super().__init__(name, age)
        self.salary = salary
        self.department = department
    def __str__(self):
        return super().__str__() + \
               'salary: ' + str(self.salary) + '\n' + \
               'department: ' + self.department + '\n'

x = Human('Jack', 24)
print(x)

y = Employee('Robert', 26, 4000, 'IT')
print(y)

#   name: Jack
#   age: 24

#   name: Robert
#   age: 26
#   salary: 4000
#   department: IT


class Human:
    def __init__(self, name, age):
        self.name = name
        self.age = age
    def __str__(self):
        return 'name: ' + self.name + '\n' + \
               'age: ' + str(self.age) + '\n'

class Employee(Human):
    def __init__(self, name, age, salary, department):
        super().__init__(name, age)
        self.salary = salary
        self.department = department
    def __str__(self):
        return super().__str__() + \
               'salary: ' + str(self.salary) + '\n' + \
               'department: ' + self.department + '\n'

x = Employee('Robert', 26, 4000, 'IT')
print(x.name, x.age, x.salary, x.department)

#    Robert 26 4000 IT


class Human:
    def __init__(self, name, age):
        self._name = name
        self._age = age
    def __str__(self):
        return 'name: ' + self.name + '\n' + \
               'age: ' + str(self.age) + '\n'

class Employee(Human):
    def __init__(self, name, age, salary, department):
        super().__init__(name, age)
        self.salary = salary
        self.department = department
    def __str__(self):
        return super().__str__() + \
               'salary: ' + str(self.salary) + '\n' + \
               'department: ' + self.department + '\n'

x = Employee('Robert', 26, 4000, 'IT')
print(x._name, x._age, x.salary, x.department)

#    Robert 26 4000 IT


class Human:
    def __init__(self, name, age):
        self.__name = name
        self.__age = age
    def __str__(self):
        return 'name: ' + self.name + '\n' + \
               'age: ' + str(self.age) + '\n'

class Employee(Human):
    def __init__(self, name, age, salary, department):
        super().__init__(name, age)
        self.salary = salary
        self.department = department
    def __str__(self):
        return super().__str__() + \
               'salary: ' + str(self.salary) + '\n' + \
               'department: ' + self.department + '\n'

x = Employee('Robert', 26, 4000, 'IT')
print(x.__name, x.__age, x.salary, x.department)

    Traceback (most recent call last):
      File "main.py", line 20, in <module>
        print(x.__name, x.__age, x.salary, x.department)
    AttributeError: 'Employee' object has no attribute '__name'

    Traceback (most recent call last):
      File "main.py", line 20, in <module>
        print(x.__age, x.salary, x.department)
    AttributeError: 'Employee' object has no attribute '__age'


class Parent:
    def __init__(self, name):
        self.parent_name = name
    def get_name(self):
        return self.parent_name

class Child(Parent):
    def __init__(self, first_name, last_name):
        super().__init__(last_name)
        self.child_name = first_name
    def get_name(self):
        return self.child_name + ' ' + self.parent_name

x = Child('Mark', 'Bill')
print(x.get_name())

#    Mark Bill


class Human:
    def __init__(self, name):
        self.name = name
    def greeting(self):
        return 'Hi my name is ' + self.name + '\n'

class Employee(Human):
    def __init__(self, name, salary):
        super().__init__(name)
        self.salary = salary

x = Employee('Mark', 10000)
print(x.greeting())

#    Hi my name is Mark


class BaseClass:
    pass
class DerivedClass(BaseClass):
    pass

class BaseClass1:
    pass
class BaseClass2:
    pass
class MultiDerived(BaseClass1, BaseClass2): 
    pass

class BaseClass:
    pass
class ChildClass(BaseClass):
    pass
class GrandChild(ChildClass):
    pass

class BaseClass:
    pass
class DerivedClass1(BaseClass):
    pass
class DerivedClass2(BaseClass):
    pass
class DerivedClass3(BaseClass):
    pass

class BaseClass:
    pass
class DerivedClass1(BaseClass):
    pass
class DerivedClass2(BaseClass):
    pass
class MultiDerived(DerivedClass1, DerivedClass2): 
    pass


class Human:
    def __init__(self, name, age):
        self.name = name
        self.age = age
    def __str__(self):
        return 'name: ' + self.name + '\n' + \
               'age: ' + str(self.age) + '\n'

class Employee(Human):
    def __init__(self, name, age, salary, department):
        Human.__init__(self, name, age)
        self.salary = salary
        self.department = department
    def __str__(self):
        return Human.__str__(self) + \
               'salary: ' + str(self.salary) + '\n' + \
               'department: ' + self.department + '\n'

x = Employee('Alex', 32, 10000, 'HR')
print(x)

#    name: Alex
#    age: 32
#    salary: 10000
#    department: HR


class Father:
    def __init__(self, name):
        self.father_name = name
    def __str__(self):
        return 'name: ' + self.father_name + '\n'

class Mother:
    def __init__(self, name):
        self.mother_name = name
    def __str__(self):
        return 'name: ' + self.mother_name + '\n'

class Child(Father, Mother):
    def __init__(self, father_name, mother_name, name):
        Father.__init__(self, father_name)
        Mother.__init__(self, mother_name)
        self.name = name
    def __str__(self):
        return Father.__str__(self) + \
               Mother.__str__(self) + \
               'name: ' + self.name + '\n'

x = Child('bill', 'alex', 'adam')
print(x)

#    name: bill
#    name: alex
#    name: adam


class Human:
    def __init__(self, name, age):
        self.name = name
        self.age = age
    def __str__(self):
        return 'name: ' + self.name + '\n' + \
               'age: ' + str(self.age) + '\n'

class Employee(Human):
    def __init__(self, name, age, salary, department):
        Human.__init__(self, name, age)
        self.salary = salary
        self.department = department
    def __str__(self):
        return Human.__str__(self) + \
               'salary: ' + str(self.salary) + '\n' + \
               'department: ' + self.department + '\n'

class Leader(Employee):
    def __init__(self, name, age, salary, department, reports):
        Employee.__init__(self, name, age, salary, department)
        self.reports = reports
    def __str__(self):
        return Employee.__str__(self) + \
               'reports: ' + str(self.reports) + '\n' 

x = Leader('Manuel', 28, 20000, 'Research', ['R1', 'R2', 'R3'])
print(x)

#    name: Manuel
#    age: 28
#    salary: 20000
#    department: Research
#    reports: ['R1', 'R2', 'R3']


class Room:
    def __init__(self, name, area):
        self.name = name
        self.area = area
    def __str__(self):
        return 'Room: ' + \
               'name: ' + self.name + ', ' + \
               'area: ' + str(self.area) + '\n'
    def __del__(self):
        del self.name
        del self.area
        print('Room deleted')

class Salesman:
    def __init__(self, name, salary):
        self.name = name
        self.salary = salary
    def __str__(self):
        return 'Salesman: ' + \
               'name: ' + self.name + ', ' + \
               'salary: ' + str(self.salary) + '\n'

class Shop:
    def __init__(self, rooms, salesmen):
        self.rooms = rooms
        self.salesmen = salesmen
    def __str__(self):
        res = ''
        for i in self.rooms: res += str(i)
        for i in self.salesmen: res += str(i)
        return res
    def __del__(self):
        for i in self.rooms: 
            Room.__del__(i)
        del self.rooms
        print('Shop deleted\n')
        

room1 = Room('R1', 60)
room2 = Room('R2', 100)
room3 = Room('R3', 80)
room4 = Room('R4', 40)
room5 = Room('R5', 70)
salesman1 = Salesman('Jack', 6000)
salesman2 = Salesman('Alex', 10000)
salesman3 = Salesman('Bill', 4000)
salesman4 = Salesman('Muller', 8000)
salesman5 = Salesman('Peter', 7000)

shop1_rooms = [room1, room2, room3]
shop1_salesmen = [salesman1, salesman2, salesman3]
shop1 = Shop(shop1_rooms, shop1_salesmen)
print(shop1)

shop2_rooms = [room4, room5]
shop2_salesmen = [salesman4, salesman5]
shop2 = Shop(shop2_rooms, shop2_salesmen)
print(shop2)

del shop1

room6 = Room('R6', 80)
room7 = Room('R7', 40)
room8 = Room('R8', 70)
shop3_rooms = [room6, room7, room8]
shop3_salesmen = [salesman1, salesman2, salesman3]
shop3 = Shop(shop3_rooms, shop3_salesmen)
print(shop3)

#    Room: name: R1, area: 60
#    Room: name: R2, area: 100
#    Room: name: R3, area: 80
#    Salesman: name: Jack, salary: 6000
#    Salesman: name: Alex, salary: 10000
#    Salesman: name: Bill, salary: 4000

#    Room: name: R4, area: 40
#    Room: name: R5, area: 70
#    Salesman: name: Muller, salary: 8000
#    Salesman: name: Peter, salary: 7000

#    Room deleted
#    Room deleted
#    Room deleted
#    Shop deleted

#    Room: name: R6, area: 80
#    Room: name: R7, area: 40
#    Room: name: R8, area: 70
#    Salesman: name: Jack, salary: 6000
#    Salesman: name: Alex, salary: 10000
#    Salesman: name: Bill, salary: 4000

File: Lectures/Object-Oriented/04-Inheritance-and-Function-Overriding.cpp
#include <bits/stdc++.h>
using namespace std;

class BaseClass {
};
class DerivedClass : public BaseClass {
};


class Human {
protected:
	string name;
	int age;
public:
	Human(string name, int age) : name(name), age(age) {}
    string display_object() {
        return "name: " + this->name + '\n' +
               "age: " + to_string(this->age) + '\n';
	}
};
class Employee : public Human {
public:
	string department;
	int salary;
public:
    Employee(string name, int age, int salary, string department) 
		: Human(name, age), salary(salary), department(department) {}
    string display_object() {
        return Human::display_object() + 
               "salary: " + to_string(this->salary) + '\n' +
               "department: " + this->department + '\n';
	}
};
int main() {
	Human x("Jack", 24);
	cout << x.display_object() << '\n';

	Employee y("Robert", 26, 4000, "IT");
	cout << y.display_object() << '\n';
}

//	name: Jack
//	age: 24

//	name: Robert
//	age: 26
//	salary: 4000
//	department: IT

class Human {
public:
	string name;
	int age;
public:
	Human(string name, int age) : name(name), age(age) {}
    string display_object() {
        return "name: " + this->name + '\n' +
               "age: " + to_string(this->age) + '\n';
	}
};
class Employee : public Human {
public:
	string department;
	int salary;
public:
    Employee(string name, int age, int salary, string department) 
		: Human(name, age), salary(salary), department(department) {}
    string display_object() {
        return Human::display_object() + 
               "salary: " + to_string(this->salary) + '\n' +
               "department: " + this->department + '\n';
	}
};
int main() {
	Employee x("Robert", 26, 4000, "IT");
	cout << x.name << ' ' << x.age << ' ' << x.salary << ' ' << x.department << '\n';
}

//    Robert 26 4000 IT

class Human {
protected:
	string name;
	int age;
public:
	Human(string name, int age) : name(name), age(age) {}
    string display_object() {
        return "name: " + this->name + '\n' +
               "age: " + to_string(this->age) + '\n';
	}
};
class Employee : public Human {
public:
	string department;
	int salary;
public:
    Employee(string name, int age, int salary, string department) 
		: Human(name, age), salary(salary), department(department) {}
    string display_object() {
        return Human::display_object() + 
               "salary: " + to_string(this->salary) + '\n' +
               "department: " + this->department + '\n';
	}
};
int main() {
	Employee x("Robert", 26, 4000, "IT");
	cout << x.name << ' ' << x.age << ' ' << x.salary << ' ' << x.department << '\n';
}

	error: 'name' is a protected member of 'Human'
			cout << x.name << ' ' << x.age << ' ' << x.salar...
					  ^
	note: declared protected here
			string name;
				   ^
	error: 'age' is a protected member of 'Human'
			cout << x.name << ' ' << x.age << ' ' << x.salar...
									   ^
	note: declared protected here
			int age;


class Human {
private:
	string name;
	int age;
public:
	Human(string name, int age) : name(name), age(age) {}
    string display_object() {
        return "name: " + this->name + '\n' +
               "age: " + to_string(this->age) + '\n';
	}
};
class Employee : public Human {
public:
	string department;
	int salary;
public:
    Employee(string name, int age, int salary, string department) 
		: Human(name, age), salary(salary), department(department) {}
    string display_object() {
        return Human::display_object() + 
               "salary: " + to_string(this->salary) + '\n' +
               "department: " + this->department + '\n';
	}
};
int main() {
	Employee x("Robert", 26, 4000, "IT");
	cout << x.name << ' ' << x.age << ' ' << x.salary << ' ' << x.department << '\n';
}

	error: 'name' is a private member of 'Human'
			cout << x.name << ' ' << x.age << ' ' << x.salar...
					  ^
	note: declared private here
			string name;
				   ^
	error: 'age' is a private member of 'Human'
			cout << x.name << ' ' << x.age << ' ' << x.salar...
									   ^
	note: declared private here
			int age;

class Parent {
protected:
	string parent_name;
public:
	Parent(string name) : parent_name(name) {}
    string get_name() {
        return this->parent_name;
	}
};
class Child : public Parent {
public:
	string child_name;
public:
    Child(string first_name, string last_name) 
		: Parent(last_name), child_name(first_name) {}
    string get_name() {
        return this->child_name + ' ' + this->parent_name;
	}
};
int main() {
	Child x("Mark", "Bill");
	cout << x.get_name() << '\n';
}

//    Mark Bill


class Human {
protected:
	string name;
public:
	Human(string name) : name(name) {}
    string greeting() {
        return "Hi my name is " + this->name + '\n';
	}
};
class Employee : public Human {
public:
	int salary;
public:
    Employee(string name, int salary) 
		: Human(name), salary(salary) {}
};
int main() {
	Employee x("Mark", 10000);
	cout << x.greeting() << '\n';
}

//    Hi my name is Mark


class BaseClass {
};
class DerivedClass : public BaseClass {
};

class BaseClass1 {
};
class BaseClass2 {
};
class MultiDerived : public BaseClass1, public BaseClass2 {
};

class BaseClass {
};
class ChildClass : public BaseClass {
};
class GrandChild : public ChildClass {
};

class BaseClass {
};
class DerivedClass1 : public BaseClass {
};
class DerivedClass2 : public BaseClass {
};
class DerivedClass3 : public BaseClass {
};

class BaseClass {
};
class DerivedClass1 : public BaseClass {
};
class DerivedClass2 : public BaseClass {
};
class MultiDerived : public DerivedClass1, public DerivedClass2 {
};


class Human {
protected:
	string name;
	int age;
public:
	Human(string name, int age) : name(name), age(age) {}
    string display_object() {
        return "name: " + this->name + '\n' +
               "age: " + to_string(this->age) + '\n';
	}
};
class Employee : public Human {
public:
	string department;
	int salary;
public:
    Employee(string name, int age, int salary, string department) 
		: Human(name, age), salary(salary), department(department) {}
    string display_object() {
        return Human::display_object() + 
               "salary: " + to_string(this->salary) + '\n' +
               "department: " + this->department + '\n';
	}
};
int main() {
	Employee x("Alex", 32, 10000, "HR");
	cout << x.display_object() << '\n';
}

//    name: Alex
//    age: 32
//    salary: 10000
//    department: HR


class Father {
protected:
	string father_name;
public:
	Father(string name) : father_name(name) {}
    string display_object() {
        return "name: " + this->father_name + '\n';
	}
};
class Mother {
protected:
	string mother_name;
public:
	Mother(string name) : mother_name(name) {}
    string display_object() {
        return "name: " + this->mother_name + '\n';
	}
};
class Child : public Father, public Mother {
public:
	string name;
public:
    Child(string father_name, string mother_name, string name) 
		: Father(father_name), Mother(mother_name), name(name) {}
    string display_object() {
        return Father::display_object() + 
			   Mother::display_object() + 
               "name: " + this->name + '\n';
	}
};
int main() {
	Child x("bill", "alex", "adam");
	cout << x.display_object() << '\n';
}

//    name: bill
//    name: alex
//    name: adam


class Human {
protected:
	string name;
	int age;
public:
	Human(string name, int age) : name(name), age(age) {}
    string display_object() {
        return "name: " + this->name + '\n' +
               "age: " + to_string(this->age) + '\n';
	}
};
class Employee : public Human {
protected:
	string department;
	int salary;
public:
    Employee(string name, int age, int salary, string department) 
		: Human(name, age), salary(salary), department(department) {}
    string display_object() {
        return Human::display_object() + 
               "salary: " + to_string(this->salary) + '\n' +
               "department: " + this->department + '\n';
	}
};
class Leader : public Employee {
public:
	string reports[3];
public:
    Leader(string name, int age, int salary, 
		   string department, string reports[3]) 
		: Employee(name, age, salary, department) {
        for (int i = 0; i < 3; this->reports[i] = reports[i], i++);
    } 
    string display_object() {
		string str_reports = "reports: ";
		for (int i = 0; i < 3; str_reports += this->reports[i] + " ", i++);
        return Employee::display_object() + 
               str_reports + '\n';
	}
};
int main() {
	string reports[3] = {"R1", "R2", "R3"};
	Leader x("Manuel", 28, 20000, "Research", reports);
	cout << x.display_object() << '\n';
}

//    name: Manuel
//    age: 28
//    salary: 20000
//    department: Research
//    reports: R1, R2, R3


class Room {
public:
	string* name;
	float* area;
public:
	Room(string* name, float* area) : name(name), area(area) {}
    string display_object() {
        return "Room: " +
               ("name: " + *(this->name) + ", " +
                "area: " + to_string(*(this->area)) + '\n');
	}
	~Room() {
		delete name;
		delete area;
        cout << "Room deleted\n";
	}
};
class Salesman {
public:
	string name;
	float salary;
public:
	Salesman(string name, float salary) : name(name), salary(salary) {}
    string display_object() {
        return "Salesman: " +
               ("name: " + this->name + ", " +
                "salary: " + to_string(this->salary) + '\n');
	}
};
class Shop {
public:
	int n_rooms;
	Room** rooms;
	int n_salesmen;
	Salesman** salesmen;
public:
	Shop(Room** rooms, int n_rooms, Salesman** salesmen, int n_salesmen) : 
		rooms(rooms), n_rooms(n_rooms), salesmen(salesmen), n_salesmen(n_salesmen) {}
    string display_object() {
		string res = "";
		for (int i = 0; i < n_rooms; res += this->rooms[i]->display_object(), i++);
		for (int i = 0; i < n_salesmen; res += this->salesmen[i]->display_object(), i++);
		return res;
	}
	~Shop() {
		for (int i = 0; i < n_rooms; delete this->rooms[i], i++);
		delete rooms;
        cout << "Shop deleted\n\n";
	}
};
int main() {
	Room* room1 = new Room(new string("R1"), new float(60));
	Room* room2 = new Room(new string("R2"), new float(100));
	Room* room3 = new Room(new string("R3"), new float(80));
	Room* room4 = new Room(new string("R4"), new float(40));
	Room* room5 = new Room(new string("R5"), new float(70));
	Salesman* salesman1 = new Salesman("Jack", 6000);
	Salesman* salesman2 = new Salesman("Alex", 10000);
	Salesman* salesman3 = new Salesman("Bill", 4000);
	Salesman* salesman4 = new Salesman("Muller", 8000);
	Salesman* salesman5 = new Salesman("Peter", 7000);

	Shop* shop1 = new Shop(new Room*[3]{room1, room2, room3}, 3, 
						   new Salesman*[3]{salesman1, salesman2, salesman3}, 3);
	cout << shop1->display_object() << '\n';

	Shop* shop2 = new Shop(new Room*[2]{room4, room5}, 2, 
						   new Salesman*[2]{salesman4, salesman5}, 2);
	cout << shop2->display_object() << '\n';

	delete shop1;

	Room* room6 = new Room(new string("R6"), new float(80));
	Room* room7 = new Room(new string("R7"), new float(40));
	Room* room8 = new Room(new string("R8"), new float(70));
	Shop* shop3 = new Shop(new Room*[3]{room6, room7, room8}, 3, 
						   new Salesman*[3]{salesman1, salesman2, salesman3}, 3);
	cout << shop3->display_object() << '\n';
}

//    Room: name: R1, area: 60.000000
//    Room: name: R2, area: 100.000000
//    Room: name: R3, area: 80.000000
//    Salesman: name: Jack, salary: 6000.000000
//    Salesman: name: Alex, salary: 10000.000000
//    Salesman: name: Bill, salary: 4000.000000

//    Room: name: R4, area: 40.000000
//    Room: name: R5, area: 70.000000
//    Salesman: name: Muller, salary: 8000.000000
//    Salesman: name: Peter, salary: 7000.000000

//    Room deleted
//    Room deleted
//    Room deleted
//    Shop deleted

//    Room: name: R6, area: 80.000000
//    Room: name: R7, area: 40.000000
//    Room: name: R8, area: 70.000000
//    Salesman: name: Jack, salary: 6000.000000
//    Salesman: name: Alex, salary: 10000.000000
//    Salesman: name: Bill, salary: 4000.000000
